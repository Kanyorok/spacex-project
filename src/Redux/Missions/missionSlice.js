import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const API_DATA = 'https://api.spacexdata.com/v3/missions';

export const displayMissions = createAsyncThunk('missions', async () => {
  try {
    const retrievedMissions = await axios.get(API_DATA);
    const missionsDisplay = retrievedMissions.data.map((mission, index) => ({
      mission_id: mission.mission_id,
      mission_name: mission.mission_name,
      description: mission.description,
      itemNumber: index + 1,
    }));
    return missionsDisplay;
  } catch (error) {
    throw Error(error);
  }
});

const initialState = {
  missions: [],
  loading: 'idle',
};

const missionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    availableMissions: (state, action) => {
      const mission = action.payload;
      state.missions = state.missions.filter((typeMission) => typeMission.mission_name === mission);
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(displayMissions.pending, (state) => {
        state.loading = 'pending';
      })
      .addCase(displayMissions.fulfilled, (state, action) => {
        state.missions = action.payload;
        state.loading = 'Succeeded loading missions';
      })
      .addCase(displayMissions.rejected, (state) => {
        state.loading = 'failed load missions';
      });
  },
});

export default missionSlice.reducer;
