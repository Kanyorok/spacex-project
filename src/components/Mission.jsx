import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { displayMissions } from '../Redux/Missions/missionSlice';
import Specificmission from './Specificmission';

const Mission = () => {
  const { missions } = useSelector((state) => state.missions);
  const [misssionLoading, setMissionLoading] = useState(true);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(displayMissions());
    setTimeout(() => {
      setMissionLoading(false);
    }, 2000);
  }, [dispatch]);

  return (
    <div>
      {misssionLoading ? (
        <p>Loading...</p>
      ) : (
        missions.map((mission) => <Specificmission key={mission.mission_id} mission={mission} />)
      )}
    </div>
  );
};

export default Mission;
