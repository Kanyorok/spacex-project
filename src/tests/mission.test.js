import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Mission from '../components/Mission';
import { displayMissions } from '../Redux/Missions/missionSlice';

jest.mock('../Redux/Missions/missionSlice', () => ({
  displayMissions: jest.fn(),
}));

const mockStore = configureStore([]);
const initialState = {
  missions: {
    loading: false,
    missions: [
      {
        mission_id: '1',
        mission_name: 'Mock Mission 1',
        description: 'Mock description 1',
        reserved: true,
      },
      {
        mission_id: '2',
        mission_name: 'Mock Mission 2',
        description: 'Mock description 2',
        reserved: false,
      },
    ],
  },
};
const store = mockStore(initialState);

test('Missions component renders correctly', () => {
  displayMissions.mockReturnValueOnce({ type: 'mockAction' });
  const tree = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <Mission />
        </BrowserRouter>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
