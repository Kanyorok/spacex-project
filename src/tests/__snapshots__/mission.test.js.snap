// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions component renders correctly 1`] = `
<div
  className="max-w-6xl mx-auto"
>
  <table
    className="table-fixed border dark:border-neutral-500 mt-4 mb-4"
  >
    <thead
      className="border-b font-medium dark:border-neutral-500"
    >
      <tr>
        <th
          className="border-r px-6 py-4 dark:border-neutral-500"
          scope="col"
        >
          Mission
        </th>
        <th
          className="border-r px-6 py-4 dark:border-neutral-500"
          scope="col"
        >
          Description
        </th>
        <th
          className="border-r px-6 py-4 dark:border-neutral-500"
          scope="col"
        >
          Status
        </th>
        <th
          className="border-r px-6 py-4 dark:border-neutral-500"
          scope="col"
        >
          â€˜
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <p>
            Loading...
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
